File: src/libopfs.js
// Copyright 2023 Roy T. Hashimoto. All Rights Reserved.

// @ts-ignore
addToLibrary({
  $opfs_support__postset: 'opfs_support();',
  $opfs_support: function() {
    class File {
      constructor(path, flags) {
        this.path = path;
        this.flags = flags;
        this.lockState = 0; // SQLITE_LOCK_NONE
        this.lockRelease = null;
      }
    };
    File.openFiles = new Map();
    File.rootDirectory = navigator.storage.getDirectory();
    File.getPathComponents = async function(path, create) {
      try {
        const [_, directories, filename] = path.match(/[/]?(.*)[/](.*)$/);
        let directoryHandle = await File.rootDirectory;
        for (const directory of directories.split('/')) {
          if (directory) {
            directoryHandle = await directoryHandle.getDirectoryHandle(directory, { create });
          }
        }
        return [directoryHandle, filename];
      } catch (e) {
        return [];
      }
    }
    File.u64 = function(ptr) {
      const index = ptr >> 2;
      return HEAPU32[index] + (HEAPU32[index + 1] * (2**32));
    }

    // @ts-ignore
    opfs_support = function() {
      return { File };
    }
  },

  opfsOpen: async function(_, zName, file, flags, pOutFlags) {
    const path = zName ? UTF8ToString(zName) : `/null_${Math.random().toString(36).slice(2)}`;
    console.log('xOpen', path, file, '0x'+flags.toString(16));
    try {
      // @ts-ignore
      const { File } = opfs_support();
      const f = new File(path, flags);
      const create = !!(flags & 0x4);
      const [directoryHandle, filename] = await File.getPathComponents(f.path, create);
      if (!directoryHandle) return 14; // SQLITE_CANTOPEN

      const fileHandle = await directoryHandle.getFileHandle(filename, { create });
      f.accessHandle = await fileHandle.createSyncAccessHandle({ mode: 'readwrite-unsafe' });
      File.openFiles.set(file, f);

      setValue(pOutFlags, flags, 'i32');
      return 0; // SQLITE_OK
    } catch (e) {
      switch (e.name) {
        case 'NotFoundError':
          return 14; // SQLITE_CANTOPEN
        default:
          return 10; // SQLITE_IOERR
      }
    }
  },
  opfsOpen__deps: ['$opfs_support'],
  opfsOpen__sig: 'ipppip',
  opfsOpen__async: true,

  opfsDelete: async function(_, zName, syncDir) {
    const path = UTF8ToString(zName);
    console.log('xDelete', path, syncDir);

    // @ts-ignore
    const { File } = opfs_support();
    const [directoryHandle, filename] = await File.getPathComponents(path, false);
    await directoryHandle.removeEntry(filename);
    return 0; // SQLITE_OK
  },
  opfsDelete__deps: ['$opfs_support'],
  opfsDelete__sig: 'ippi',
  opfsDelete__async: true,

  opfsAccess: async function(_, zName, flags, pResOut) {
    const path = UTF8ToString(zName);
    console.log('xAccess', path, '0x'+flags.toString(16));

    try {
      // @ts-ignore
      const { File } = opfs_support();
      const [directoryHandle, filename] = await File.getPathComponents(path, false);
      await directoryHandle.getFileHandle(filename);
      setValue(pResOut, 1, 'i32');
    } catch (e) {
      if (e.name === 'NotFoundError') {
        setValue(pResOut, 0, 'i32');
      } else {
        return 10; // SQLITE_IOERR
      }
    }
    return 0; // SQLITE_OK
  },
  opfsAccess__deps: ['$opfs_support'],
  opfsAccess__sig: 'ippip',
  opfsAccess__async: true,

  opfsClose: async function(file) {
    // @ts-ignore
    const { File } = opfs_support();
    const f = File.openFiles.get(file);
    console.log('xClose', f?.path);
    f.accessHandle.close();
    File.openFiles.delete(file);

    if (f.flags & 8) {
      // SQLITE_OPEN_DELETEONCLOSE
      const [directoryHandle, filename] = await File.getPathComponents(f.path, false);
      await directoryHandle.removeEntry(filename);
    }
    return 0; // SQLITE_OK
  },
  opfsClose__deps: ['$opfs_support'],
  opfsClose__sig: 'ip',
  opfsClose__async: true,

  opfsRead: function(file, pData, iAmt, pOffset) {
    // @ts-ignore
    const { File } = opfs_support();
    const f = File.openFiles.get(file);
    const iOffset = File.u64(pOffset);
    console.log('xRead', f?.path, iAmt, iOffset);

    const pDataArray = HEAPU8.subarray(pData, pData + iAmt);
    const nBytes = f.accessHandle.read(pDataArray, { at: iOffset });
    if (nBytes !== iAmt) return 522; // SQLITE_IOERR_SHORT_READ
    return 0; // SQLITE_OK
  },
  opfsRead__deps: ['$opfs_support'],
  opfsRead__sig: 'ippii',

  opfsWrite: function(file, pData, iAmt, pOffset) {
    // @ts-ignore
    const { File } = opfs_support();
    const f = File.openFiles.get(file);
    const iOffset = File.u64(pOffset);
    console.log('xWrite', f?.path, iAmt, iOffset);

    const pDataArray = HEAPU8.subarray(pData, pData + iAmt);
    const nBytes = f.accessHandle.write(pDataArray, { at: iOffset });
    if (nBytes !== iAmt) return 778; // SQLITE_IOERR_WRITE
    return 0; // SQLITE_OK
  },
  opfsWrite__deps: ['$opfs_support'],
  opfsWrite__sig: 'ippii',

  opfsTruncate: function(file, pSize) {
    // @ts-ignore
    const { File } = opfs_support();
    const f = File.openFiles.get(file);
    const iSize = File.u64(pSize);
    console.log('xTruncate', f?.path, iSize);
    
    f.accessHandle.truncate(iSize);
    return 0; // SQLITE_OK
  },
  opfsTruncate__deps: ['$opfs_support'],
  opfsTruncate__sig: 'ipp',

  opfsSync: function(file, flags) {
    // @ts-ignore
    const { File } = opfs_support();
    const f = File.openFiles.get(file);
    console.log('xSync', f?.path, '0x'+flags.toString(16));

    f.accessHandle.flush();
    return 0; // SQLITE_OK
  },
  opfsSync__deps: ['$opfs_support'],
  opfsSync__sig: 'ipi',

  opfsFileSize: function(file, pSize) {
    // @ts-ignore
    const { File } = opfs_support();
    const f = File.openFiles.get(file);
    const dv = new DataView(HEAPU8.buffer, pSize, 8);
    const size = f.accessHandle.getSize();
    dv.setBigInt64(0, BigInt(size), true);
    console.log('xFileSize', f.path, size);
    return 0; // SQLITE_OK
  },
  opfsFileSize__deps: ['$opfs_support'],
  opfsFileSize__sig: 'ipp',

  opfsLock: async function(file, flags) {
    // @ts-ignore
    const { File } = opfs_support();
    const f = File.openFiles.get(file);
    console.log('xLock', f?.path, '0x'+flags.toString(16));

    if (flags && f.lockState === 0) {
      await new Promise(acquired => {
        navigator.locks.request(f.path, lock => {
          acquired();
          return new Promise(releaser => {
            f.lockRelease = releaser;
          });
        });
      });
    }
    f.lockState = flags;

    return 0; // SQLITE_OK
  },
  opfsLock__deps: ['$opfs_support'],
  opfsLock__sig: 'ipi',
  opfsLock__async: true,

  opfsUnlock: function(file, flags) {
    // @ts-ignore
    const { File } = opfs_support();
    const f = File.openFiles.get(file);
    console.log('xUnlock', f?.path, '0x'+flags.toString(16));

    if (!flags && f.lockState !== 0) {
      f.lockRelease();
      f.lockRelease = null;
    }
    f.lockState = flags;

    return 0; // SQLITE_OK
  },
  opfsUnlock__deps: ['$opfs_support'],
  opfsUnlock__sig: 'ipi',

  opfsCheckReservedLock: async function(file, pResOut) {
    // @ts-ignore
    const { File } = opfs_support();
    const f = File.openFiles.get(file);
    console.log('xCheckReservedLock', f?.path);

    setValue(pResOut, 0, 'i32');
    return 0; // SQLITE_OK
  },
  opfsCheckReservedLock__deps: ['$opfs_support'],
  opfsCheckReservedLock__sig: 'ipp',
  opfsCheckReservedLock__async: true,

  opfsFileControl: async function(file, flags, pOut) {
    // @ts-ignore
    const { File } = opfs_support();
    const f = File.openFiles.get(file);
    console.log('xFileControl', f?.path, flags);
    return 12; // SQLITE_NOTFOUND
  },
  opfsFileControl__deps: ['$opfs_support'],
  opfsFileControl__sig: 'ipip',
  opfsFileControl__async: true,

  opfsSectorSize: function(file) {
    // @ts-ignore
    const { File } = opfs_support();
    const f = File.openFiles.get(file);
    console.log('xSectorSize', f?.path);
    return 4096; // SQLITE_OK
  },
  opfsSectorSize__deps: ['$opfs_support'],
  opfsSectorSize__sig: 'ip',

  opfsDeviceCharacteristics: function(file) {
    // @ts-ignore
    const { File } = opfs_support();
    const f = File.openFiles.get(file);
    console.log('xDeviceCharacteristics', f?.path);
    return 2048; // SQLITE_IOCAP_UNDELETABLE_WHEN_OPEN
  },
  opfsDeviceCharacteristics__deps: ['$opfs_support'],
  opfsDeviceCharacteristics__sig: 'ip',

});

File: src/libopfs.c
// Copyright 2023 Roy T. Hashimoto. All Rights Reserved.
#include <sys/time.h>
#include <emscripten.h>
#include <sqlite3.h>
#include <string.h>

// sqlite3_io_methods javascript handlers
// 64-bit integer parameters are passed by pointer.
extern int opfsClose(sqlite3_file* file);
extern int opfsRead(sqlite3_file* file, void* pData, int iAmt, const sqlite3_int64* pOffset);
extern int opfsWrite(sqlite3_file* file, const void* pData, int iAmt, const sqlite3_int64* pOffset);
extern int opfsTruncate(sqlite3_file* file, const sqlite3_int64* pSize);
extern int opfsSync(sqlite3_file* file, int flags);
extern int opfsFileSize(sqlite3_file* file, sqlite3_int64* pSize);
extern int opfsLock(sqlite3_file* file, int flags);
extern int opfsUnlock(sqlite3_file* file, int flags);
extern int opfsCheckReservedLock(sqlite3_file* file, int* pResOut);
extern int opfsFileControl(sqlite3_file* file, int flags, void* pOut);
extern int opfsSectorSize(sqlite3_file* file);
extern int opfsDeviceCharacteristics(sqlite3_file* file);

extern int opfsOpen(sqlite3_vfs* vfs, const char *zName, sqlite3_file* file, int flags, int *pOutFlags);
extern int opfsDelete(sqlite3_vfs* vfs, const char *zName, int syncDir);
extern int opfsAccess(sqlite3_vfs* vfs, const char *zName, int flags, int *pResOut);

// Glue functions to pass 64-bit integers by pointer.
static int xRead(sqlite3_file* file, void* pData, int iAmt, sqlite3_int64 iOffset) {
  return opfsRead(file, pData, iAmt, &iOffset);
}
static int xWrite(sqlite3_file* file, const void* pData, int iAmt, sqlite3_int64 iOffset) {
  return opfsWrite(file, pData, iAmt, &iOffset);
}
static int xTruncate(sqlite3_file* file, sqlite3_int64 size) {
  return opfsTruncate(file, &size);
}
static int xOpen(sqlite3_vfs* vfs, const char* zName, sqlite3_file* file, int flags, int* pOutFlags) {
  static sqlite3_io_methods io_methods = {
    1,
    opfsClose,
    xRead,
    xWrite,
    xTruncate,
    opfsSync,
    opfsFileSize,
    opfsLock,
    opfsUnlock,
    opfsCheckReservedLock,
    opfsFileControl,
    opfsSectorSize,
    opfsDeviceCharacteristics
  };
  file->pMethods = &io_methods;

  return opfsOpen(vfs, zName, file, flags, pOutFlags);
}

static int xFullPathname(sqlite3_vfs* vfs, const char* zName, int nOut, char* zOut) {
  strncpy(zOut, zName, nOut);
  return SQLITE_OK;
}

static int xCurrentTime(sqlite3_vfs* vfs, double* pJulianDay) {
  // UNIX epoch 1/1/1970 is Julian day 2440587.5
  static const sqlite3_int64 unixEpoch = 24405875*(sqlite3_int64)8640000;
  struct timeval sNow;
  gettimeofday(&sNow, 0);
  sqlite3_int64 julianMillis = unixEpoch + 1000*(sqlite3_int64)sNow.tv_sec + sNow.tv_usec/1000;
  *pJulianDay = julianMillis / 86400000.0;
  return SQLITE_OK;
}

void* EMSCRIPTEN_KEEPALIVE getSqliteFree() {
  return sqlite3_free;
}

int main() {
  sqlite3_initialize();

  sqlite3_vfs* vfs = (sqlite3_vfs*)sqlite3_malloc(sizeof(sqlite3_vfs));
  vfs->iVersion = 1;
  vfs->szOsFile = sizeof(sqlite3_file);
  vfs->mxPathname = 1024;
  vfs->pNext = NULL;
  vfs->zName = "opfs";
  vfs->pAppData = NULL;
  vfs->xOpen = xOpen;
  vfs->xDelete = opfsDelete;
  vfs->xAccess = opfsAccess;
  vfs->xFullPathname = xFullPathname;
  vfs->xCurrentTime = xCurrentTime;
  
  // Get remaining functionality from the default VFS.
  sqlite3_vfs* defer = sqlite3_vfs_find(0);
#define COPY_FIELD(NAME) vfs->NAME = defer->NAME
  COPY_FIELD(xDlOpen);
  COPY_FIELD(xDlError);
  COPY_FIELD(xDlSym);
  COPY_FIELD(xDlClose);
  COPY_FIELD(xRandomness);
  COPY_FIELD(xSleep);
  COPY_FIELD(xGetLastError);
#undef COPY_FIELD

  sqlite3_vfs_register(vfs, 0);

  return 0;
}
